### PR TITLE
CMake cleanup, added License and all menu files

### PR DESCRIPTION
**Pull request by stefonarch**
Can be improved more I guess.
Provides file to fix https://github.com/lxqt/lxqt-menu-data/issues/2

### CODE REVIEW

**Comment by yan12125** [comment]
More CMake stuffs can be removed. I started a branch https://github.com/lxqt/lxqt-menu-data/tree/cmake-cleanup some time ago but forgot about it.

**Comment by tsujan** [comment]
IMO, there can't be any "unneeded stuff" in the app chooser dialog.

Logically, we can't assume anything about any item. Consider the example of Qps. Just because it can't open a file, we shouldn't remove it, for two reasons: (1) We can't know how Qps works only based on its desktop entry; and (2) We might make it open a kind of file for whatever reason in future. There are so many other examples too.

Those 2 reason apply to all apps. Moreover, the absence of `MimeType` in the desktop entry doesn't show that the app can't open any file. For example, XTerm can "open" any executable bash script.

As another example, a "preferences" app may be able to open a "preferences" file with a special extension to load it, to edit it, to verify it, etc. There are so many possibilities...

In short, *we shouldn't assume anything about any app*, not even our apps (because we might change them in future).

**Comment by tsujan** [comment]
Oh, and don't forget that LXQt's App chooser (in File Association dialog) doesn't exclude anything in the "Other applications" section either: you can *try* to associate  Session Settings to an audio MIME type with it  — perhaps for the same reasons I mentioned above.

**Comment by stefonarch** [comment]
> In short, we shouldn't assume anything about any app, not even our apps (because we might change them in future).

If we will change _ever_ ` lxqt-about`, screenlocker or leave menu items to open  some files we could  insert them in the `applications-only.menu` again...

For "Preferences" I think basically the same: it never happened and probably never will that any app is associated with a mimetype or can open a file.  

IMO a menu should'n contain any useless categories just because in a eventual improbable future they could be needed.

**Comment by tsujan** [comment]
> IMO a menu should'n contain any useless categories just because in a eventual improbable future they could be needed.

I added Kvantum Manager to the LXQt category several years ago, when I used KDE, and I didn't need a permission from LXQt devs (whom I joined later).

Since we have no control over the contents of categories, we shouldn't assume anything about them, let alone considering them "useless".

**Comment by yan12125** [comment]
On one hand, I agree with tsujan that LXQt shouldn't exclude unusual/unlikely use cases, and on the other hand, I agree with stef - it's indeed strange to see "Leave", "About LXQt" and "Lock Screen" like the screenshot in #2.

A compromise may be adding a new `.menu` file with Preferences submenu and without "Leave", "About LXQt" and "Lock Screen". That may be achieved by copying https://github.com/lxde/lxmenu-data/blob/master/layout/lxde-applications.menu and replacing `lxde-*.directory` with `lxqt-*.directory`.

I haven't tried it out yet - just a thought.

**Comment by tsujan** [comment]
> A compromise may be adding

That's still questionable. Apart from making a simple change complicated for no good reason:

There are — or can be — a lot of other items like "Leave", "About LXQt",..., depending on which apps are installed. Why should we treat our items differently?

And what's the goal? To prevent users from selecting "About LXQt" for an audio MIME type? Such safeguards aren't justified. We can't prevent users from selecting so many other meaningless items and, unlike GNOME, LXQt doesn't see its users stupid.

Or is the goal making the list "uncluttered" by removing only a few items among so many senseless ones?

Sorry, but from whatever angle I look at it, I can't justify it.

**Comment by stefonarch** [comment]
The question is simple: we never showed the leave menu, lxqt-about and screensaver in this menu.

If somebody comes up and says "Hey, this menu is not really complete, I will make a PR to add the above items" we would merge it ?!?

**Comment by tsujan** [comment]
> we never showed the leave menu, lxqt-about and screensaver in this menu.

I think we did nothing; they just couldn't be in a tree that belonged to LXDE.

I'll hesitantly agree with excluding them — but not the settings category — if they don't have a side effect for `menu://applications/` after libfm-qt is patched correctly.

**Comment by stefonarch** [comment]
I realize now that the old menu didn't have "Others", Leave items are packed into "system tools"


> > we never showed the leave menu, lxqt-about and screensaver in this menu.
> 
> I think we did nothing; they just couldn't be in a tree that belonged to LXDE.
> 


> ~I'll hesitantly agree with excluding them — but not the settings category — if they don't have a side effect for `menu://applications/` after libfm-qt is patched correctly.~
> 

There is no side effect, see my screenshot.

> EDIT: The tab name is "Installed Applications".

I have nothing against showing all and every applications in their categories - but not 3 main categories/items which hardly/never are used.

> A compromise may be adding a new .menu file with Preferences submenu and without "Leave", "About LXQt" and "Lock Screen". That may be achieved by copying https://github.com/lxde/lxmenu-data/blob/master/layout/lxde-applications.menu and replacing lxde-*.directory with lxqt-*.directory.


I agree on this, things will be  like before and we get rid of all lxde-*stuff.

**Comment by tsujan** [comment]
I looked into the related part of libfm-qt's code and started to see the whole picture. Please bare with me and tell me if something is unclear in what follows:

 1. `lxqt-applications.menu` belongs to lxqt-panel. Therefore, we can't use it in lbfm-qt (we can't expect users to have lxqt-panel to use pcmanfm-qt). Is there a plan to include `lxqt-applications.menu` in this source and make lxqt-panel dependent on it?

 2. Or is the plan to use this package *only* with libfm-qt? If yes, should `lxqt-applications-only.menu` be used by libfm-qt?

 3. More importantly, the app chooser dialog isn't the only place where libfm-qt uses `lxde-applications.menu`. It also uses that file for  `menu://applications/` **if `XDG_MENU_PREFIX` isn't set.** Of course, it's set under LXQt (which means `menu://applications/` loads lxqt-panel's `lxqt-applications.menu` under LXQt), but **if it isn't set outside LXQt**, either (1) we should use  the same file that we use for the app chooser dialog, in which case, *changing any category/item in that file will affect `menu://applications/` too*, (e.g.. if you remove "LXQt About" from that file, it'll also be removed from `menu://applications/`), or (2) we need another file for `menu://applications/`.

What's the plan regarding 1, 2 and 3?

**Comment by tsujan** [comment]
> There is no side effect, see my screenshot.

Please read my above comment → 3.

**Comment by stefonarch** [comment]
The plan is to replace the  `lxmenu-data` dependency with one for `lxqt-menu-data` which provides category names and the menu file used in `libfm-qt` (2). There shouldn't be any difference if we use the same menu file just replacing lxde→lxqt; with `lxqt-panel` installed the categories and items are the same as in mainmenu, without `lxqt-applications.menu` and no prefix set it will look like before.

For 1) it's the same, but it's easier to let the (main)menu files where they are now (there inserted in weblate).

**Comment by tsujan** [comment]
I wasn't an LXDE user but the comments and code lines of the file `vfs-menu.c` suggest that the reason behind `lxmenu-data` was that both libfm and lxpanel  needed it. Now, if only libfm-qt is going to use `lxqt-menu-data`, then the latter could be included in the former. Why creating a separate package?

Including it in libfm-qt has another advantage too: If a user has a libfm-qt based app like lximage-qt, *without having pcmanfm-qt*, it'll be guaranteed that the app chooser dialog works for him. Currently, there's no guarantee because he may not have `lxmenu-data`.

**Comment by stefonarch** [comment]
>  Now, if only libfm-qt is going to need lxqt-menu-data, then the latter could be included in the former. Why creating a separate package?

The panel needs it too, for the category translations in the main menu. 
If we include all in `libfm-qt` the panel has to depend on it.

**Comment by tsujan** [comment]
The panel doesn't *need* it, because it has its own `lxqt-applications.menu`. Of course, it can just *use* it when available (almost always).

**Comment by tsujan** [comment]
OK, let's do it as @yan12125 suggested, i.e., using `lxde-applications.menu` as the model and replacing `lxde-*.directory` with `lxqt-*.directory`. After that, `lxqt-applications.menu` should be moved to `lxqt-menu-data` from lxqt-panel; otherwise it would be incomplete.

**Comment by stefonarch** [comment]
OK, let's do it as @yan12125 suggested, i.e., using lxde-applications.menu as the model and replacing lxde-*.directory with lxqt-*.directory

I'll add it here, I think `appchooser.menu` is a better name to make it clearer for what it's used.

> After that, lxqt-applications.menu should be moved to lxqt-menu-data from lxqt-panel; otherwise it would be incomplete.

You mean all 3 menu files actually provided by lxqt-panel/menu?

**Comment by tsujan** [comment]
> I think appchooser.menu is a better name

It'll also be used by `menu://applications/` where `XDG_MENU_PREFIX` isn't set (→ https://github.com/lxqt/lxqt-menu-data/pull/3#issuecomment-1520523296 → 3). It's usable with lxqt-panel's main menu too. I think the previous name you gave to it was better.

> You mean all 3 menu files actually provided by lxqt-panel/menu?

Since `lxqt-menu-data` is supposed to be the menu data of LXQt, yes, it should contain all of them. But I don't know how difficult it'll be to transfer the current translations.

**Comment by stefonarch** [comment]
Translations are not hard to transfer, and never ever got updates. I'm not sure if it's worth the pain to enable them on the other  directory files, it would need some scripting (I had it but I forgot all about).

The previous name was correct only for a menu without "Settings". But a name can be found.

**Comment by tsujan** [comment]
> The previous name was correct only for a menu without "Settings". But a name can be found.

What about `lxde-applications-fm.menu`?

**Comment by yan12125** [comment]
> This one exists: https://specifications.freedesktop.org/menu-spec/menu-spec-1.0

That page mentions:

> Menu files for this version of the specification must use the following namespace, public and system identifiers: 
> Namespace
>      http://www.freedesktop.org/standards/menu
> Public Identifier for 1.0
>     PUBLIC "-//freedesktop//DTD Menu 1.0//EN" 
> System Identifier for 1.0
>    http://www.freedesktop.org/standards/menu-spec/menu-1.0.dtd

Probably freedesktop.org should fix the standard or their website. I suggest to keep the link mentioned in the standard for now.

**Comment by stefonarch** [comment]
I'm  not  sure if moving all 3 other menu files and eventually also `desktop-directories` from lxqt-panel to here is a good idea. 
Users who just install pcmanfm-qt  will end up with all our menus while only one is needed (maybe therefor also drop "lxqt-" from` lxqt-applications-fm.menu` ?)

**Comment by tsujan** [comment]
As its name says, it's LXQt's menu data. It isn't just for pcmanfm-qt, like its LXDE counterpart that wasn't only for pcmanfm. All menus LXQt uses will be in it.

The package will be small. Even adding tens of menus to it won't make a difference.

**Comment by stefonarch** [comment]
It's not about the package, it's about filling up non-lxqt users'  `/etc/xdg/menus/`  folder. 
I wondered many time what this `lxde-applications.menu` was for as I never had LXDE.

**Comment by tsujan** [comment]
Those small files can't be a problem to anyone. Compare it to what other DEs do. We are so considerate that we prefix our file names with `lxqt-`. KDE puts a file named `applications.menu` there, as if that folder belongs to it.

BTW, I think  it's quite logical to put `lxqt-config.menu` in `lxqt-menu-data` too.

**Comment by stefonarch** [comment]
I updated the PR; the PR for lxqt-panel has to be updated too if we agree on those migrations.

> BTW, I think it's quite logical to put lxqt-config.menu in lxqt-menu-data too.

At this point why not, but that will make depend `lxqt-config` on this here too.

**Comment by tsujan** [comment]
> that will make depend lxqt-config on this here too.

Since `lxqt-config` and `lxqt-panel` are LXQt apps, it won't be a surprise.

As for libfm-qt/pcmanfm-qt, they always needed `lxmenu-data`.

What worries me is that, since it's a run-time dependency, some people may not install it and think there's a bug — although this possibility has always existed for pcmanfm-qt. Is there a way to make it impossible?

**Comment by stefonarch** [comment]
>Is there a way to make it impossible?

Maybe require it in CMake (it's not my field) like 

`find_package(lxqt-build-tools ${LXQTBT_MINIMUM_VERSION} REQUIRED)`

**Comment by tsujan** [comment]
> Maybe require it in CMake

It's possible and I like it but, since it isn't a build dependency, I'm not sure if others agree with it.

**Comment by stefonarch** [comment]
Another way is let them stay where they are and keep 'lxqt-menu-data` exactly like `lxmenu-data` was...

**Comment by tsujan** [comment]
> ...keep lxqt-menu-data exactly like lxmenu-data was

I didn't understand your point. `lxqt-menu-data` already follows `lxmenu-data`.

This problem isn't new, and it may not be seen as a problem by many. For example, `lxqt-panel` needs `lxmenu-data` as a run-time dependency, but it doesn't depend on it. Logically, it should be done by ditros, as in https://archlinux.org/packages/community/x86_64/lxqt-panel/

My argument is that this kind of dependency is too important, at least for `lxqt-panel` if not for pcmanfm-qt.

**Comment by yan12125** [comment]
(From https://github.com/lxqt/lxqt-menu-data/pull/3#issuecomment-1520523296)

> 3. More importantly, the app chooser dialog isn't the only place where libfm-qt uses `lxde-applications.menu`. It also uses that file for  `menu://applications/` **if `XDG_MENU_PREFIX` isn't set.** Of course, it's set under LXQt (which means `menu://applications/` loads lxqt-panel's `lxqt-applications.menu` under LXQt), but **if it isn't set outside LXQt**, either (1) we should use  the same file that we use for the app chooser dialog, in which case, _changing any category/item in that file will affect `menu://applications/` too_, (e.g.. if you remove "LXQt About" from that file, it'll also be removed from `menu://applications/`), or (2) we need another file for `menu://applications/`.
> 

I haven't had time to look into details. Let me check if I'm looking at the same things as you. By "the app chooser dialog" and `menu://applications/`, do you mean https://github.com/lxqt/libfm-qt/blob/1.3.0/src/appmenuview.cpp#L43 and https://github.com/lxqt/libfm-qt/blob/1.3.0/src/core/vfs/vfs-menu.c#L247 ?

> It's possible and I like it but, since it isn't a build dependency, I'm not sure if others agree with it.

It's good to put runtime dependencies in CMakeLists.txt. That can notify not only users who build LXQt components from sources but also distro package maintainers - unfornately, some packagers do not look at upstream changelogs and ship packages without testing it.

An example is https://github.com/lxqt/lxqt-session/pull/428 (Add qtxdg-tools as a dependency). I think there are more in LXQt.

**Comment by tsujan** [comment]
> By "the app chooser dialog" and menu://applications/, do you mean...

Yes, but for `menu://applications/,` it's also in the synchronous lookup: https://github.com/lxqt/libfm-qt/blob/954c016ce0f8f1bade1000b82f76c23df9e2d5de/src/core/vfs/vfs-menu.c#L1094
Anyway, @stefonarch and I reached an agreement, based on your suggestion.

> It's good to put runtime dependencies in CMakeLists.txt

Good to know that you agree! Putting such dependencies in `CMakeLists.txt` can save users a headache, especially when they compile from sources. It'll be very good if we add `lxqt-menu-data` to the cmake files of all components that needs it, including `libfm-qt` too.

**Comment by stefonarch** [comment]
> > ...keep lxqt-menu-data exactly like lxmenu-data was
> 
> I didn't understand your point. `lxqt-menu-data` already follows `lxmenu-data`.
> 

I meant that it is now more than `lxmenu-data` was, as it is containing all the application menus and `desktop-direcories` files from `lxqt-panel`. But if we add it to the relatives CMake.lists there's no risk.

**Comment by stefonarch** [comment]
@yan12125 Something important is removed in your `CMake.list.txt` -it installs the menu files to `/menu´ now:
```
-- Installing: /usr/share/desktop-directories/lxqt-utility-accessibility.directory
-- Installing: /usr/share/desktop-directories/lxqt-utility.directory
-- Installing: /menus/lxqt-applications.menu
-- Installing: /menus/lxqt-applications-compact.menu
-- Installing: /menus/lxqt-applications-simple.menu
-- Installing: /menus/lxqt-applications-fm.menu

```

**Comment by yan12125** [comment]
> stefonarch and I reached an agreement, based on your suggestion.

Great! I think I catch up the discussion. It's good to move all menu-related files to here.

> the PR for lxqt-panel has to be updated too if we agree on those migrations.

Yes, I will create a new PR after this one is merged.

> I meant that it is now more than lxmenu-data was, as it is containing all the application menus and desktop-direcories files from lxqt-panel. But if we add it to the relatives CMake.lists there's no risk.

Probably adding more comments to menu files, README files, package descriptions, ... can mitigate confusion from users.

> Something important is removed in your CMake.list.txt -it installs the menu files to `/menu´ now:

Good catch! Sorry I don't test my changes first. Here a better fix to avoid dependency on liblxqt:

```Diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 46dcee4..db340b5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,8 +11,8 @@ include(GNUInstallDirs)
 set(LXQTBT_MINIMUM_VERSION "0.12.0")
 
 find_package(lxqt-build-tools ${LXQTBT_MINIMUM_VERSION} REQUIRED)
-find_package(lxqt ${LXQT_MINIMUM_VERSION} REQUIRED)
 
+include(LXQtConfigVars)
 include(LXQtPreventInSourceBuilds)
 include(LXQtTranslate)
 
```

`LXQtConfigVars` is needed as it defines `${LXQT_ETC_XDG_DIR}` used in installation paths.

**Comment by stefonarch** [comment]
Added some file info in the README and fixed the CMake.list.txt

> It's good to move all menu-related files to here.

I looked at the [files](https://github.com/lxqt/lxqt-config/tree/master/src/menuname) in `lxqt-config`, it needs a careful hand and some of the existing here could be dropped, I'm too tired now  but it's not straightforward.

* "lxqt-settins-lxqt" and "lxqt-settings"  have the same content
* "lxqt-settings-system" are present in both, but in this repo here it is "Amministration" (which can be dropped IMO)

And there are 2 "other": "lxqt-other.directory" and "lxqt-settings-other.directory", one for the menu (Other) one for "Other Settings" in `lxqt-config`.

**Comment by yan12125** [comment]
Thanks! This PR now looks good. Other fixes/improvements can be done in future PRs.

**Review by yan12125 — APPROVED** [review]

**Comment by stefonarch** [comment]
Ok, I've done most of the migrations from `lxqt-config` files now, will make a new PR for that.
